[FLINK-9964][table] Add a full RFC-compliant CSV table format factory

twalthr opened the pull request:
## What is the purpose of the change

This PR adds a RFC-4180 standard compliant CSV table format. 

## Brief change log

It adds:

- `CsvRowSerializationSchema`
- `CsvRowDeserializationSchema`
- A new `CSV` descriptor
- `CsvRowTableFormatFactory`
- documentation and tests

This PR further improves #6541 to make the CSV format robust and configurable enough to serve most use cases. It integrates nicely with most SQL types.

It deprecates the "old CSV" descriptor stack and prepares also for FLINK-7050 (#4660).

The old CSV descriptor is still available under as "OldCsv".

## Verifying this change

Tests are added for descriptor, factory, and de/serialization schema.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

dawidwys commented:
@flinkbot approve description
@flinkbot approve consensus

dawidwys commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvRowDeSerializationSchemaTest.java:
```diff
@@ -0,0 +1,223 @@
+		deserializationSchema.setIgnoreParseErrors(true);
+
+		// unsupported null for integer
+		assertEquals(Row.of("Test", null, "Test"), deserializationSchema.deserialize("Test,null,Test".getBytes()));
```
nit: Could we have those as separate tests? Personally prefer this approach as it immediately shows what failed (by the test name).

twalthr commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java:
```diff
@@ -0,0 +1,268 @@
+		}
+	}
+
+	private Row convertRow(JsonNode node, RowTypeInfo info, boolean mapByName) {
```
This is due to a weird Jackson behavior. The index getter does not work.

dawidwys replied:
Right, got it. :( That's unfortunate...

twalthr commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:
```diff
@@ -0,0 +1,259 @@
+	// --------------------------------------------------------------------------------------------
+
+	private void convertRow(ObjectNode reuse, Row row, RowTypeInfo rowTypeInfo) {
+		if (reuse == null) {
```
Good point :)

twalthr commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:
```diff
@@ -0,0 +1,259 @@
+		}
+	}
+
+	private ArrayNode convertRow(Row row, RowTypeInfo rowTypeInfo) {
```
Must not necessarily be nested.

twalthr commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java:
```diff
@@ -0,0 +1,268 @@
+		return typeInfo;
+	}
+
+	public void setFieldDelimiter(char c) {
```
I agree.

twalthr commented on flink-formats/flink-csv/pom.xml:
```diff
@@ -0,0 +1,123 @@
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-test-utils-junit</artifactId>
```
Very good point. Nobody noticed it so far.

dawidwys replied:
The same applies to `flink-json`.

dawidwys commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:
```diff
@@ -169,10 +194,7 @@ public int hashCode() {
-		if (reuse == null) {
-			reuse = csvMapper.createObjectNode();
-		}
+	private void convertNestedRow(ObjectNode reuse, Row row, RowTypeInfo rowTypeInfo) {
```
Shouldn't this be just `convertRow`?

twalthr replied:
Sorry, my IDE was too eager.

dawidwys commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:
```diff
@@ -240,19 +260,23 @@ private JsonNode convert(ContainerNode<?> container, Object obj, TypeInformation
+	private ArrayNode convertNestedRow(Row row, RowTypeInfo rowTypeInfo) {
 		final ArrayNode arrayNode = csvMapper.createArrayNode();
 		final TypeInformation[] types = rowTypeInfo.getFieldTypes();
+		if (row.getArity() != types.length) {
```
Extract to helper method? The same block is used twice.

dawidwys commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvRowDeSerializationSchemaTest.java:
```diff
@@ -90,20 +90,20 @@ public void testSerializeDeserialize() throws IOException {
-			serSchema.setArrayElementDelimiter(":");
-			serSchema.setFieldDelimiter(';');
+		final Consumer<CsvRowSerializationSchema.Builder> serConfig = (serSchemaBuilder) -> {
+			serSchemaBuilder.setEscapeCharacter('*');
```
Use the fluent API?

dawidwys commented on flink-formats/flink-csv/src/test/java/org/apache/flink/table/descriptors/CsvTest.java:
```diff
@@ -43,6 +43,20 @@
 		)}
 	);
 
+	private static final Descriptor DESCRIPTOR_1 = new Csv()
```
`DESCRIPTOR_1` -> `CUSTOM_DESCRIPTOR_WITH_MANUAL_SCHEMA`
`DESCRIPTOR_2` -> `DEFAULT_DESCRIPTOR_WITH_DERIVED_SCHEMA`

dawidwys commented on flink-formats/flink-csv/src/test/java/org/apache/flink/table/descriptors/CsvTest.java:
```diff
@@ -56,27 +70,17 @@ public void testMissingSchema() {
 		// we add an additional schema
-		addPropertyAndVerify(descriptors().get(1), "format.schema", "DDD");
+		addPropertyAndVerify(
+			descriptors().get(1),
```
Use `DEFAULT_DESCRIPTOR_WITH_DERIVED_SCHEMA`?

dawidwys replied:
The same above.

twalthr commented:
Thanks for the review @dawidwys. I forgot to add the JAR to the SQL Client e2e test. I will do this while merging and also quickly very the format manually in the CLI.

dawidwys commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java:
```diff
@@ -209,91 +213,157 @@ public int hashCode() {
+		Object convert(JsonNode node);
+	}
+
+	private static RuntimeConverter createRowRuntimeConverter(
```
I would split this method into sth like:

	private static RuntimeConverter createRowRuntimeConverter(
			RowTypeInfo rowTypeInfo,
			boolean ignoreParseErrors,
			boolean isTopLevel) {
		final TypeInformation<?>[] fieldTypes = rowTypeInfo.getFieldTypes();
		final String[] fieldNames = rowTypeInfo.getFieldNames();

		final RuntimeConverter[] fieldConverters = createFieldConverters(ignoreParseErrors, fieldTypes);

		// Jackson only supports mapping by name in the first level
		return assembleRowConverter(ignoreParseErrors, isTopLevel, fieldNames, fieldConverters);
	}

dawidwys commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java:
```diff
@@ -209,91 +213,157 @@ public int hashCode() {
+			final RuntimeConverter fieldConverter =
+				createNullableRuntimeConverter(fieldType, ignoreParseErrors);
+
+			fieldConverters[i] = (node) -> {
```
Why not just `fieldConverters[i] = createNullableRuntimeConverter(fieldType, ignoreParseErrors);` ?

twalthr replied:
We could simplify the logic a bit. I agree. I will put the checking into the method.

dawidwys commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:
```diff
@@ -194,91 +199,121 @@ public int hashCode() {
+		JsonNode convert(CsvMapper csvMapper, ContainerNode<?> container, Object obj);
+	}
+
+	private static RuntimeConverter createRowRuntimeConverter(RowTypeInfo rowTypeInfo, boolean isTopLevel) {
```
Split similarly as in `Deserialization` schema

dawidwys commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:
```diff
@@ -71,9 +75,10 @@
 
 	private CsvRowSerializationSchema(
-			TypeInformation<Row> typeInfo,
+			RowTypeInfo typeInfo,
```
Like this change! ;)

dawidwys commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java:
```diff
@@ -209,91 +213,157 @@ public int hashCode() {
+
+		// Jackson only supports mapping by name in the first level
+		if (isTopLevel) {
+			return (node) -> {
```
Could we extract common class for those branches? The only difference is how we extract children of `JsonNode`.

dawidwys commented:
@flinkbot approve architecture

twalthr commented:
Thanks for the review @dawidwys. I addressed your comments. I think we don't need an end-to-end test for everything. Esp. with the serialization test we should be good now. If you think we should have those nevertheless. Feel free to open an follow-up issue.

Will merge it now due to feature freeze.
@flinkbot approve quality
